src/librustc_trans/allocator.rs

```diff
@@ -0,0 +1,117 @@
+                AllocatorTy::Bang |
+                AllocatorTy::ResultExcess |
+                AllocatorTy::ResultPtr |
+                AllocatorTy::ResultUnit |
+                AllocatorTy::UsizePair |
+                AllocatorTy::Unit => panic!("invalid allocator arg"),
+            }
+        }
+        let output = match method.output {
+            AllocatorTy::UsizePair => {
+                args.push(usizep); // min
+                args.push(usizep); // max
+                None
+            }
+            AllocatorTy::Bang => None,
+            AllocatorTy::ResultExcess => {
+                args.push(i8p); // excess_ptr
+                args.push(i8p); // err_ptr
+                Some(i8p)
+            }
+            AllocatorTy::ResultPtr => {
+                args.push(i8p); // err_ptr
+                Some(i8p)
+            }
+            AllocatorTy::ResultUnit => Some(i8),
+            AllocatorTy::Unit => None,
+
+            AllocatorTy::AllocErr |
+            AllocatorTy::Layout |
+            AllocatorTy::LayoutRef |
+            AllocatorTy::Ptr => panic!("invalid allocator output"),
+        };
+        let ty = llvm::LLVMFunctionType(output.unwrap_or(void),
+                                        args.as_ptr(),
+                                        args.len() as c_uint,
+                                        False);
+        let name = CString::new(format!("__rust_{}", method.name)).unwrap();
+        let llfn = llvm::LLVMRustGetOrInsertFunction(llmod,
+                                                     name.as_ptr(),
+                                                     ty);
+
+        let callee = CString::new(kind.fn_name(method.name)).unwrap();
+        let callee = llvm::LLVMRustGetOrInsertFunction(llmod,
+                                                       callee.as_ptr(),
+                                                       ty);
+
+        let llbb = llvm::LLVMAppendBasicBlockInContext(llcx,
+                                                       llfn,
+                                                       "entry\0".as_ptr() as *const _);
+
+        let llbuilder = llvm::LLVMCreateBuilderInContext(llcx);
+        llvm::LLVMPositionBuilderAtEnd(llbuilder, llbb);
+        let args = args.iter().enumerate().map(|(i, _)| {
+            llvm::LLVMGetParam(llfn, i as c_uint)
+        }).collect::<Vec<_>>();
+        let ret = llvm::LLVMRustBuildCall(llbuilder,
+                                          callee,
+                                          args.as_ptr(),
+                                          args.len() as c_uint,
+                                          ptr::null_mut(),
+                                          "\0".as_ptr() as *const _);
+        llvm::LLVMSetTailCall(ret, True);
+        if output.is_some() {
+            llvm::LLVMBuildRet(llbuilder, ret);
+        } else {
+            llvm::LLVMBuildRetVoid(llbuilder);
+        }
+        llvm::LLVMDisposeBuilder(llbuilder);
```

Comment:
Can't you use the `Builder` API? Also, why is there both this and a syntax expansion step?